apollo-gradle-plugin/src/main/java/com/apollographql/android/gradle/ApolloIRGenTask.java

```diff
@@ -1,172 +1,173 @@
+      String irOutput = outputDir.getAbsolutePath() + "/" + getProject().relativePath(entry.getValue().getSchemaFile().getParent());
+      new File(irOutput).mkdirs();
+
+      List<String> apolloArgs = Lists.newArrayList("generate");
+      apolloArgs.addAll(entry.getValue().getQueryFiles());
+      apolloArgs.addAll(Lists.newArrayList("--schema", entry.getValue().getSchemaFile().getAbsolutePath(),
+          "--output", irOutput + "/" + Utils.capitalize(variant) + "API.json", "--target", "json"));
+      setArgs(apolloArgs);
+      super.exec();
     }
-
-    setScript(apolloScript);
-    List<String> apolloArgs = Lists.newArrayList("generate");
-    Set<String> inputPathSet = Sets.newHashSet(Iterables.transform(getInputFiles(), new Function<File, String>() {
-      @Nullable
-      @Override
-      public String apply(@Nullable File file) {
-        return getProject().file(file).getAbsolutePath();
-      }
-    }));
-    apolloArgs.addAll(inputPathSet);
-    apolloArgs.addAll(Lists.newArrayList("--schema", schemaFile.getAbsolutePath(),
-        "--output", outputDir.getAbsolutePath() + "/" + Utils.capitalize(variant) + "API.json",
-        "--target", "json"));
-    setArgs(apolloArgs);
-    super.exec();
   }
 
-  private File userProvidedSchemaFile() {
-    File schemaFile = null;
-    ImmutableList<String> schemaFiles = FluentIterable.from(config)
-        .transform(new Function<GraphQLExtension, String>() {
-          @Nullable
-          @Override
-          public String apply(@Nullable GraphQLExtension graphQLExtension) {
-            return (graphQLExtension != null ? graphQLExtension.getSchemaFile() : null);
-          }
-        })
-        .filter(Predicates.notNull())
-        .toList();
-
-    if (!schemaFiles.isEmpty()) {
-      if (schemaFiles.size() > 1) {
-        throw new IllegalArgumentException("More than two schema files were specified for the build variant "
-            + variant + ". Please ensure that only one schema field is specified for $variant's source sets");
+  /**
+   * Extracts schema files from the task inputs and sorts them in a way similar to the Gradle lookup priority.
+   * That is, build variant source set, build type source set, product flavor source set and finally main
+   * source set.
+   *
+   * The schema file under the source set with the highest priority is used and all the graphql query files under the
+   * schema file's subdirectories from all source sets are used to generate the IR.
+   *
+   * If any of the schema file's ancestor directories contain a schema file, a GradleException is
+   * thrown. This is considered to be an ambiguous case.
+   *
+   * @param files - task input files which consist of .graphql query files and schema.json files
+   * @return - a map with schema files as a key and associated query files as a value
+   */
+  private Map<String, ApolloCodegenArgs> buildSchemaQueryMap(Set<File> files) {
```

Comment:
There seems to quite a bit of logic in this method. Maybe it would make sense to break it up a bit and unit test some pieces?